server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -1396,4 +1309,139 @@ private static Instant getTimestampFromParser(BytesReference source, XContentTyp
     public static Instant getCanonicalTimestampBound(Instant time) {
         return time.truncatedTo(ChronoUnit.SECONDS);
     }
+
+    public static class Builder {
+        private LongSupplier timeProvider;
+        private String name;
+        private List<Index> indices;
+        private long generation;
+        private Map<String, Object> metadata;
+        private boolean hidden;
+        private boolean replicated;
+        private boolean system;
+        private boolean allowCustomRouting;
+        private IndexMode indexMode;
+        @Nullable
+        private DataStreamLifecycle lifecycle;
+        private boolean rolloverOnWrite;
+        private boolean failureStore;
+        private List<Index> failureIndices;
+        @Nullable
+        private DataStreamAutoShardingEvent autoShardingEvent;
+
+        public Builder() {}
```

Comment:
I would check out how something like IndexMetadata works with its builder - it has required parameters in the constructor if not copying an existing instance.

Reply:
I think I was a bit quick with adding the empty constructor here (force of habit). On second thought (or maybe first thought 🙃), it might actually make sense to only have the "copy" constructor for now, as we don't have any other use cases currently for using a builder to construct a `DataStream` from scratch. We don't have a lot of optional fields and, looking at the usages of the `DataStream` constructors, it doesn't seem to add any value to have defaults for fields. Would you be okay with removing this constructor altogether? That would also resolve your other question on fields not being set.

Reply:
I think the main things we would need to always include would be the name and the first write index. Everything else could have a sensible default based on the current creation logic:

```
name: Required
indices: Required

generation: Default 1L
metadata: Default null
hidden: Default false
replicated: Default false
system: Default false
allowCustomRouting: Default false
indexMode: Default null
lifecycle: Default null
failureStore: Default false
failureIndices: Default empty list
rolloverOnWrite: Default false
autoShardingEvent: Default null
```